ngraph/python/tests/test_onnx/utils/model_zoo_tester.py

```diff
@@ -0,0 +1,143 @@
+        # type: (Type[Backend], List[Dict[str,str]], Optional[str]) -> None
+        self.backend = backend
+        self._parent_module = parent_module
+        self._include_patterns = set()  # type: Set[Pattern[Text]]
+        self._exclude_patterns = set()  # type: Set[Pattern[Text]]
+        self._test_items = defaultdict(dict)  # type: Dict[Text, Dict[Text, TestItem]]
+
+        for zoo_model in zoo_models:
+            test_name = "test_{}".format(zoo_model["model_name"])
+
+            test_case = OnnxTestCase(
+                name=test_name,
+                url=zoo_model["url"],
+                model_name=zoo_model["model_name"],
+                model_dir=None,
+                model=None,
+                data_sets=None,
+                kind="OnnxBackendRealModelTest",
+                rtol=zoo_model.get("rtol", 0.001),
+                atol=zoo_model.get("atol", 1e-07),
+            )
+            self._add_model_test(test_case, "Zoo")
+
+    @staticmethod
+    @retry
+    def _get_etag_for_url(url):  # type: (str) -> str
+        request = urlopen(url)
+        return request.info().get("ETag")
+
+    @staticmethod
+    def _read_etag_file(model_dir):  # type: (str) -> str
+        etag_file_path = os.path.join(model_dir, "source_tar_etag")
+        if os.path.exists(etag_file_path):
+            return open(etag_file_path).read()
+
+    @staticmethod
+    def _write_etag_file(model_dir, etag_value):  # type: (str, str) -> None
+        etag_file_path = os.path.join(model_dir, "source_tar_etag")
+        open(etag_file_path, "w").write(etag_value)
+
+    @staticmethod
+    def _backup_old_version(model_dir):  # type: (str) -> None
+        if os.path.exists(model_dir):
+            backup_index = 0
+            while True:
+                dest = "{}.old.{}".format(model_dir, backup_index)
+                if os.path.exists(dest):
+                    backup_index += 1
+                    continue
+                shutil.move(model_dir, dest)
+                break
+
+    @classmethod
+    @retry
+    def prepare_model_data(cls, model_test):  # type: (OnnxTestCase) -> Text
+        onnx_home = os.path.expanduser(os.getenv("ONNX_HOME", os.path.join("~", ".onnx")))
+        models_dir = os.getenv("ONNX_MODELS", os.path.join(onnx_home, "models"))
+        model_dir = os.path.join(models_dir, model_test.model_name)  # type: Text
+        current_version_etag = ModelZooTestRunner._get_etag_for_url(model_test.url)
+
+        # If model already exists, check if it's the latest version by verifying cached Etag value
+        if os.path.exists(os.path.join(model_dir, "model.onnx")):
+            if not current_version_etag or current_version_etag == ModelZooTestRunner._read_etag_file(
+                model_dir
+            ):
+                return model_dir
+
+            # If model does exist, but is not current, backup directory
+            ModelZooTestRunner._backup_old_version(model_dir)
```

Comment:
Making backup of all old models can choke the CI machines in the long run if these models are saved.
Maybe this should be configurable and disabled by default?

Reply:
I think backups should be extended - the old model should be restored if obtaining a new version fails. In case of a success, the old model should be deleted.

Reply:
Good idea. I'll create a story for this.